moped-editor/src/views/projects/projectView/KnackSync.js

```diff
@@ -0,0 +1,217 @@
+      knackApplicationId = "6167314778435d001ea3e7cb",
+    },
+    ref
+  ) => {
+    /**
+     * Function to build the correct Knack URL to interact with based on properties and if there will be an
+     * update or an initial sync.
+     * @returns string
+     */
+    const buildUrl = () => {
+      let url =
+        "https://api.knack.com/v1/pages/scene_" +
+        sceneNumber +
+        "/views/view_" +
+        viewNumber +
+        "/records";
+      if (project.moped_project[0].knack_project_id) {
+        // existing record
+        url = url + "/" + project.moped_project[0].knack_project_id;
+      }
+      return url;
+    };
+
+    /**
+     * Function to determine the HTTP method to use base on if there will be an update or initial post to Knack
+     * @returns string
+     */
+    const getHttpMethod = () => {
+      let method = "POST";
+      if (project.moped_project[0].knack_project_id) {
+        method = "PUT";
+      }
+      return method;
+    };
+
+    /**
+     * Function to build up the headers which need to be sent as part of the a Knack API call
+     * @returns object
+     */
+    const buildHeaders = () => {
+      let headers = {
+        "Content-Type": "application/json",
+        "X-Knack-Application-Id": knackApplicationId,
+        "X-Knack-REST-API-Key": "knack",
+      };
+      return headers;
+    };
+
+    /**
+     * Function to build up a JSON object of the fields which need to be updated in a call to Knack. This is needed
+     * because if you update the project number field, even with the same, extant number, Knack returns an error.
+     * @returns string
+     */
+    const buildBody = () => {
+      let body = {};
+
+      let field_map = {
+        field_3998: "project_id",
+        field_3999: "project_name",
+        field_4000: "current_status",
+      };
+
+      Object.keys(field_map).forEach(element => {
+        if (
+          project.moped_project[0].currentKnackState[element] !==
+          project.moped_project[0][field_map[element]]
+        ) {
+          body[element] = project.moped_project[0][field_map[element]];
+        }
+      });
+
+      return JSON.stringify(body);
+    };
+
+    const [mutateProjectKnackId] = useMutation(UPDATE_PROJECT_KNACK_ID);
+
+    /**
+     * Function to hanlde the actual mechanics of syncronizing the data on hand to the Knack API endpoint.
+     */
+    const handleSync = () => {
+      if (project.moped_project[0].knack_project_id) {
+        // updating knack record
+        fetch(buildUrl(), {
+          method: "GET",
+          headers: buildHeaders(),
+        })
+          .then(response => response.json())
+          .then(
+            result => {
+              if (result.errors) {
+                // knack error
+                snackbarHandler({
+                  severity: "warning",
+                  message: "Success: Data Tracker sync failed.",
+                });
+                return Promise.reject(result);
```

Comment:
Use `.catch()` instead of `.then(..., onRejected)`, onRejected never handles rejected promises from the same then callback, you are also prone to have race conditions if the chain is misqueued.

This implementation is not necessarily bad, it simply is cryptic to read and difficult to debug because the trace of Promise.reject gets lost in the chain.

Reply:
@frankhereford [here's](https://github.com/cityofaustin/atd-moped/blob/3f5539935dec8a93dd7163ff69d0adb9ff2037a5/moped-editor/src/utils/mapHelpers.js#L1060) a `.catch()` example for you.

Reply:
It took me a while to understand what you were pointing out, but, yes - this now makes perfect sense to me. Thank you for pointing this out, and I agree fully.